### PR TITLE
Initial Coding Standard Proposal

### PR DESCRIPTION
**Pull request by RyanBerge**
This document is quite large and I found myself skimming through it near the end.  Everything in there is open to discussion.  There are a few places in particular where I've added some comments requesting opinions.  Let me know if anything seems weird or if there's something I should have removed but missed.

When you have time, please take the opportunity to review the standard in its entirety; some of these points you will likely be well-familiar with, but some may introduce you to concepts you haven't seen before.

### CODE REVIEW

**Comment by collinsc** [comment]
Can it be programmed into a style script?

**Comment by Jebbs** [comment]
Good suggestion. Checking as much of the coding standard as possible during CI would be pretty cool.

File: CodingStandard.md [hunk 1]
```diff
@@ -0,0 +1,1425 @@
...
+referenced outside of their compilation unit must be declared with internal
+linkage.
+
+Variables (and functions and types) declared and used within a single
+compilation unit should be declared inside an anonymous namespace, instead of
+using the `static` keyword.
+
+**Casting**
+
+C++ style casts must be used. Other cast formats (such as C-style
```
Comment (collinsc) [inline CodingStandard.md:110]: Is this static_cast and dynamic_cast only?

File: CodingStandard.md [hunk 2]
```diff
@@ -0,0 +1,1425 @@
...
+
+
+**Memory**
+
+A function must not return a pointer or reference to memory allocated
+within that function with temporary or automatic storage duration.
+
+Classes that allocate memory with dynamic storage duration, including
+any memory allocated with the new operator, or allocate other shared resources
+must explicitly define copy constructors and copy assignment operators.
```
Comment (collinsc) [inline CodingStandard.md]: Copy assignment for shared resource? why not shared_ptr?

Comment (collinsc) [reply CodingStandard.md]: or even better, not have shared resources.

File: CodingStandard.md [hunk 3]
```diff
@@ -0,0 +1,1425 @@
...
+"is-a" relationship and subtypes (publicly derived classes) must be
+substitutable.
+
+Use of multiple inheritance or virtual inheritance (that is, using the virtual
+keyword in the inheritance specification for a class to make one or more of its
+parent classes be virtual base classes) requires justification.
+
+**Complexity**
+
+Code functions must not be overly complex.
```
Comment (collinsc) [inline CodingStandard.md]: There is a potential for more specific wording here.

File: CodingStandard.md [hunk 4]
```diff
@@ -0,0 +1,1425 @@
...
+compatible with the target hardware and software architecture.
+
+**Unused Return Values**
+
+The return value of a non-void function must be explicitly cast to (void)
+using a C-style cast if the value is not used by the calling function.
+
+[I believe the purpose here is to show that anyone ignoring the return value 
+of a function is doing so intentionally, but I'm considering pulling it anyway
+because it's a little weird.  Thoughts?]
```
Comment (collinsc) [inline CodingStandard.md]: I have never seen that before, but I think it is meant to encourage using error codes:

if (!file.open()){
  //handle error here
}

File: CodingStandard.md [hunk 5]
```diff
@@ -0,0 +1,1425 @@
...
+
+*C++ exceptions must not be used.*
+
+[This was more of a hardware limitation for UASC, but it may make sense for us as 
+
+well. 
+Read up on the
+<a href="https://en.wikipedia.org/wiki/Design_by_contract">Design By Contract</a>
+programming style (possibly to be included as a separate document).]
+
```
Comment (collinsc) [inline CodingStandard.md]: Not agreeing with you yet. Does this rule out sections of the std not marked noexcept? If this is identified as a priority then it needs to be reworded as end user code or our code. I don't think the ros libraries make liberal use of noexcept so that would be impossible to enforce.

File: CodingStandard.md [hunk 6]
```diff
@@ -0,0 +1,1425 @@
...
+**dynamic_cast**
+
+Except for dynamic_cast, use of Run-Time Type Identification (RTTI) such as
+the typeid operator requires justification.
+
+**Friend**
+
+Using the friend keyword requires justification.
```
Comment (collinsc) [inline CodingStandard.md:291]: I dislike friend, it better be good justificaiton.

File: CodingStandard.md [hunk 7]
```diff
@@ -0,0 +1,1425 @@
...
+
+**Protected**
+
+Using the protected keyword requires justification.
+
+**Initialization Lists**
+
+Constructor initializer lists must have the same order as the member
+data in the class definition.
```
Comment (collinsc) [inline CodingStandard.md:296]: Also never change order, or add declaration of member variables in class without checking order of initializer lists.

File: CodingStandard.md [hunk 8]
```diff
@@ -0,0 +1,1425 @@
...
+**Software Design Constraints**
+
+- Global data must be documented in the design.
+- Global data must be minimized.
+- Bi-directional dependencies between software elements must be justified.
+- Exceptions and assertions used for control flow must be justified. Exception:
+  a timer exception used for task scheduling.
+- The use of self-modifying software is prohibited.
+
+**Software Design Guidelines**
```
Comment (collinsc) [inline CodingStandard.md]: i think we should revisit these more abstract requirements as a team

File: CodingStandard.md [hunk 9]
```diff
@@ -0,0 +1,1425 @@
...
+  such as scheduling, interrupt and event-driven architectures, dynamic tasking,
+  re-entry, global data and exception handling.
+- The design should have high cohesion within software elements.
+- The design should have loose coupling between software elements.
+- The design should include well-defined interface descriptions between software
+  elements.
+- The design should include a hardware abstraction layer (HAL) that provides a
+  device-independent interface to hardware services.
+
+**Object-Oriented Software Design Guidelines**
```
Comment (collinsc) [inline CodingStandard.md]: We aren't using OOP as our primary paradigm

**Comment by AdamuKaapan** [comment]
I actually think ROS has a way to run a code format checker as part of build (if not by default, we could totally make it happen anyways), which would therefore also make it run during a Travis CI build. Does anyone know of any good C++ formatting checkers?

File: CodingStandard.md (hunk 2 above)
Comment (RyanBerge) [reply CodingStandard.md]: You're right---shared_ptr would be preferred. Will update.

File: CodingStandard.md (hunk 1 above)
Comment (RyanBerge) [reply CodingStandard.md:110]: I'm not certain what you mean.  

A C-style cast, such as `(int)myVar` attempts to perform several different types of casts before reporting a failure (I believe it does a `static_cast` first, then a `dynamic_cast`, then a `reinterpret_cast`, and only if all three fail does it throw an exception).  Requiring c++ casts decreases the chance of a bug due to an incorrectly programmed cast.

File: CodingStandard.md (hunk 3 above)
Comment (RyanBerge) [reply CodingStandard.md]: Agreed, but I wasn't sure what to say; this is literally all the words that were in this section of the standard I copied.

File: CodingStandard.md (hunk 4 above)
Comment (RyanBerge) [reply CodingStandard.md]: You're correct.  There are more specific examples in a different part of the standard, though I think I cut that part because it was talking about alternatives to exception handling.

```
[[nodiscard]]
int Foo() { return 3; }

int main()
{
    (void)Foo();
}
```

With C++17, the `[[nodiscard]]` tag will tell the compiler to generate a warning if the return value is ignored, *unless* the return value is *explicitly discarded*.  I think.

File: CodingStandard.md (hunk 1 above)
Comment (Jebbs) [reply CodingStandard.md:110]: I wonder if Collin is talking about explicitly limiting the casts we are allowed to use to `static_cast` and `dynamic_cast`?

File: CodingStandard.md (hunk 5 above)
Comment (RyanBerge) [reply CodingStandard.md]: True.  Perhaps the only reason to avoid exceptions for us is that they, rarely, can leave your OS in an unexpected state after having been handled.  <a href="https://randomascii.wordpress.com/2012/07/05/when-even-crashing-doesnt-work/">See this article for details</a> (and quite a bit of other interesting tidbits).

Granted, the problem he describes is with Windows 64-bit, particularly, but it's strange enough to make me wary of exceptions in general.

File: CodingStandard.md (hunk 6 above)
Comment (RyanBerge) [reply CodingStandard.md:291]: Agreed, but I didn't want to ban it completely, because I think there are, occasionally, problems that can only be solved with `friend`.

File: CodingStandard.md (hunk 8 above)
Comment (RyanBerge) [reply CodingStandard.md]: We certainly can.  They all seemed reasonable to me.

**Comment by RyanBerge** [comment]
Why can't I reply to specific people in the general conversation section?

Adam, `clang-tidy` has format-enforcing checks, and I believe it's configurable, though I haven't ever attempted to do configurations before.

**Comment by Jebbs** [comment]
@RyanBerge The most you can do is mention them (which may or may not be what you want)

File: CodingStandard.md (hunk 1 above)
Comment (collinsc) [reply CodingStandard.md:110]: Yeah I was I talking about limiting, to static_cast, but in general not subverting the type system. e.g. I don't want to see 
char c = inputMethod();
if ((int)char == 32;)
  ...

At least outside of a self contained utility method that does the conversion,.

File: CodingStandard.md [hunk 10]
```diff
@@ -0,0 +1,1425 @@
...
+- Use of interfaces as a means to decrease coupling is recommended; allow
+  clients to depend on interfaces to software elements, not on their
+  implementations.
+- Composition/aggregation is recommended instead of implementation inheritance.
+- Use of inheritance solely to provision for unlikely future product
+  improvements should be avoided. Abstracting an interface and separating it
+  from an implementation as reasonable accommodation for change may be
+  appropriate.
+- Avoid allocating too many responsibilities to a single class.
+
```
Comment (collinsc) [inline CodingStandard.md:355]: we should add a clause about doing our best to adhere to ROS constructs, and frameworks.

File: CodingStandard.md [hunk 11]
```diff
@@ -0,0 +1,1425 @@
...
+  - All source code filenames should be lower case.
+
+  - C++ files should use the .cpp and .h extensions (.hpp extension
+    should not be used).
+
+  - Header files that are #included in the body of another file (e.g. to define
+    a reusable lookup table) should use the -inline.h suffix.
+
+  - All C++ files associated with a design package should be prefixed with the
```
Comment (collinsc) [inline CodingStandard.md]: We should rename some of our packages to be briefer if we are sticking by this. I have some massively long package names

File: CodingStandard.md [hunk 12]
```diff
@@ -0,0 +1,1425 @@
...
+  - Compile-time constants (const, constexpr, enumerators, etc.): `kCamelCase`
+  - Variables and Function Parameters: `lower_case`
+  - constexpr Methods and Functions: `lower_case`
+  - Non-const struct member variables: `lower_case`
+  - Non-const class member variables: `lower_case_` (incl. trailing undercore)
+  - Const member variables: `kCamelCase`
+  - Template type parameters: `CamelCase`
+  - Template value parameters: `kCamelCase`
+  - Namespaces (4 characters max) : `lower_case`
+
```
Comment (collinsc) [inline CodingStandard.md]: Pretty sure we should find a beautifier or static analysis tool to automate this.

File: CodingStandard.md [hunk 13]
```diff
@@ -0,0 +1,1425 @@
...
+- Hungarian naming convention or any encoding of names to imply type information
+  should not be used.  This includes prefixing `p` for pointers, `m` for member
+  variables, `s` for static variables, etc.
+```
+[These are the standards we were using at UASC, and I'm not particularly attached
+to them. Specifically, I wouldn't mind having a convention that differentiates
+between public and private members and methods, and I definitely dislike the 
+trailing `_` they use for class members.  What do you guys like?]
+
```
Comment (collinsc) [inline CodingStandard.md:426]: things I dislike
m_string
my_string
myString

I prefer
longer_names_with_underscores
or 
camelTypenames

but I'll do whatever we go with.

File: CodingStandard.md [hunk 14]
```diff
@@ -0,0 +1,1425 @@
...
+some_file.h would be guarded as follows:
+
+```c++
+#ifndef SOME_FILE_H
+#define SOME_FILE_H
+
+   ...
+
+#endif // SOME_FILE_H
+```
```
Comment (collinsc) [inline CodingStandard.md:440]: I think avoid
#pragma 
anything would be good.

File: CodingStandard.md [hunk 15]
```diff
@@ -0,0 +1,1425 @@
...
+
+} // namespace foos
+
+#endif // FOOS_FOO_H
+```
+
+Sections should be ordered public, protected, then private.
+
+Class members should be declared in this order:
+ - Using-declarations, typedefs and enums
```
Comment (collinsc) [inline CodingStandard.md:515]: I dislike using, I think using explicit namespaces is clearer, and can avoid some ambiguity.

However I do like auto when appropriate.

File: CodingStandard.md [hunk 16]
```diff
@@ -0,0 +1,1425 @@
...
+
+```
+
+Member function definitions can appear in any order in the .cpp file, but
+should be ordered logically to help the reader.
+
+The #include for the class header should be the first substantive line in the
+source file, following the file header and the copyright block.
+
+Source code should be wrapped at 90 characters.
```
Comment (collinsc) [inline CodingStandard.md]: I have seen 80 for legacy, but why 90?
Also tabs or spaces?

File: CodingStandard.md [hunk 17]
```diff
@@ -0,0 +1,1425 @@
...
+**Namespaces**
+
+Namespaces should be closed with a comment as in `} // namespace foos` or
+`} // namespace` for an unnamed (anonymous) namespace.
+
+Namespace contents (named or unnamed) should not be indented.
+
+**Namespace Using Directives**
+
+Namespace using directives (eg, `using namespace fplm;`) should not be used.
```
Comment (collinsc) [inline CodingStandard.md:821]: thumbs up

File: CodingStandard.md [hunk 18]
```diff
@@ -0,0 +1,1425 @@
...
+
+**Enum Classes**
+
+The C++11 scoped `enum class` should be preferred over the unscoped C-style
+`enum`.
```
Comment (collinsc) [inline CodingStandard.md:826]: yes please

File: CodingStandard.md [hunk 19]
```diff
@@ -0,0 +1,1425 @@
...
+it helps readability, but otherwise does not change the behavior of the code.
+
+Non-local variables should not be declared using `auto`, including function
+parameters or return types, except for creating a generic template-like
+function.
+
+**Assertions**
+
+Run-time conditional checks (assertions) should be used liberally to help
+expose software bugs in development and production builds.
```
Comment (collinsc) [inline CodingStandard.md:846]: static_assert is great I love seeing static_assert

File: CodingStandard.md [hunk 20]
```diff
@@ -0,0 +1,1425 @@
...
+
+<u>Summary</u>
+
+- Public inheritance implies interface inheritance and represents an "is-a"
+  relationship
+- All publicly derived classes must meet the requirements of the base class
+- Interface inheritance must satisfy the Liskov Substitution Principle
+- Avoid mixing interface inheritance and implementation inheritance if possible
+- Consider using abstract base classes, the non-virtual interface idiom, or the
+  curiously recurring template parameter idiom
```
Comment (collinsc) [inline CodingStandard.md]: Let's talk about these idioms, I haven't heard them outside this document

File: CodingStandard.md [hunk 21]
```diff
@@ -0,0 +1,1425 @@
...
+class that is substituted at compile-time, resulting in static dispatch and
+opportunities to catch errors during compilation.
+
+The `override` keyword should be used when overriding virtual methods.  `final`
+should be used for leaf classes in a class hierarchy.  Class methods should be
+marked ether `final` or `override`, but not both.
+
+Virtual methods should not be called inside a constructor or destructor.
+
+**Multiple Inheritance**
```
Comment (collinsc) [inline CodingStandard.md:1019]: I am not the biggest fan of multiple inheritance.

File: CodingStandard.md [hunk 22]
```diff
@@ -0,0 +1,1425 @@
...
+
+When used in this way, there is no need to use the `virtual` keyword to resolve
+multiple inheritance ambiguities.
+
+**Parametric Polymorphism (Templates and Macros)**
+
+<u>Summary</u>
+
+- Use templates judiciously, to enhance readability, reduce code duplication,
+  and provide quality abstractions
```
Comment (collinsc) [inline CodingStandard.md:1032]: yes yes yes

File: CodingStandard.md [hunk 23]
```diff
@@ -0,0 +1,1425 @@
...
+<u>Summary</u>
+
+- Do not use the C++ exception standard library
+- Do not use exception-related keywords `try`, `catch`, `throw`, `noexcept`
+- Code is compiled using `-fno-exceptions` to disable exceptions
+- Define interfaces with narrow contracts and use assertions to ensure they
+  are fulfilled
+- Check all return values or explicitly discard them with a cast to `void`
+- Use `[[nodiscard]]` to ensure that return values are always checked or
+  explicitly discarded
```
Comment (collinsc) [inline CodingStandard.md:1360]: With embedded systems this is commonplace. But I think we need to define an explict uniform system of error codes, or use the ROS error codes.

**Review by collinsc — COMMENTED** [review]
I like the standard as far as standards go. It's a game of compromise, however we need to discuss it as a team and make revisions before I can give it explicit approval

File: CodingStandard.md (hunk 11 above)
Comment (RyanBerge) [reply CodingStandard.md]: I think it's definitely an optional standard.  Our packages were `snsr`, `nav`, `apex`, `flpn`, etc.

File: CodingStandard.md (hunk 12 above)
Comment (RyanBerge) [reply CodingStandard.md]: `clang-tidy`

File: CodingStandard.md (hunk 15 above)
Comment (RyanBerge) [reply CodingStandard.md:515]: I like `using` as a replacement for typedef: `using IntegerType = int32_t`.

File: CodingStandard.md (hunk 16 above)
Comment (RyanBerge) [reply CodingStandard.md]: I've always used 90.  No particular reason.

We had a server script that replaced tabs with spaces so that developers could use whichever they preferred when working.

File: CodingStandard.md (hunk 20 above)
Comment (RyanBerge) [reply CodingStandard.md]: I admit I'm not really familiar with them either.  I'll figure out what they're talking about or possibly just remove this section.

File: CodingStandard.md (hunk 13 above)
Comment (Jebbs) [reply CodingStandard.md:426]: I'm actually a big fan of m_xxx for member variables. Especially those with names that would be nice to have for parameters.

Comment (LinuxKid21) [reply CodingStandard.md:426]: I double support for m_xxx for private member variables. I think it makes the code more self documenting (no need to go into the header to determine the source of a m_xxx variable).

File: CodingStandard.md (hunk 14 above)
Comment (LinuxKid21) [reply CodingStandard.md:440]: I would disagree. I think #pragma once is much simpler than the alternative of #ifndef xxx #define xxx #endif. It has less boilerplate, less duplication (no xxx which is typed twice) and no worthless variables (we need to invent the xxx macro for the sole purpose of having xxx).

As far as #pragma once being nonstandard goes. I don't think that's a problem. It works on every compiler that we will ever touch while working on this and all foreseeable, future packages.

(Otherwise, all uses of pragma should be destroyed with fire)

**Comment by LinuxKid21** [comment]
I don't see anywhere in this that explains where we should handle our template definitions(maybe I missed it). Are we going to just put it in the header file? In an "implementation file" and include it? Other thoughts?

File: CodingStandard.md (hunk 13 above)
Comment (Jebbs) [reply CodingStandard.md:426]: Yes, private is the keyword here.

File: CodingStandard.md (hunk 1 above)
Comment (RyanBerge) [reply CodingStandard.md:110]: Technically, that would resolve to a `static_cast` :)

**Comment by RyanBerge** [comment]
@LinuxKid21 Template definitions will have to go in a header file.  The only ways I know of splitting it into two files are on a programmer-side-only basis (the compiler ends up reimplementing the definitions in everyfile that uses the templates anyway) and only seem to serve to make the code harder to read.

**Comment by AdamuKaapan** [comment]
Do we have any deadline for approving this or at least, like, is this "sooner" or "later"? This is quite a lot of stuff and I'm wondering how much of a priority it should be for processing all of it

**Comment by RyanBerge** [comment]
@AdamuKaapan This is not urgent. I'm not placing a deadline on it.

File: CodingStandard.md (hunk 3 above)
Comment (Jebbs) [reply CodingStandard.md]: Maybe a bit can be included on preferring readable/maintainable solutions over more complex/performant ones. Anything with high complexity/low readability should be heavily documented and justified.

File: CodingStandard.md [hunk 24]
```diff
@@ -0,0 +1,1425 @@
...
+production source code unless they are clearly part of the source code
+documentation (e.g. examples).
+
+Unused code is: *Compile time excluded non-executable source code. That is,
+source code that is discarded by the compilation process and does not result in
+production executable object code (or data).*
+
+**Goto**
+
+The `goto` keyword should not be used.
```
Comment (Jebbs) [inline CodingStandard.md]: I disagree that `goto` should be outlawed completely, but I think justification is needed and it should be clearly marked as being used. Basically, `goto` should be a last resort, but sometimes it can be the best solution for the problem at hand.

File: CodingStandard.md (hunk 4 above)
Comment (Jebbs) [reply CodingStandard.md]: This seemed overkill to me at first glance, but after thinking about it more I am warming up to this practice. Are we able to use `[[nodiscard]]` with our project? Surely we should be able to use it in some capacity.

File: CodingStandard.md [hunk 25]
```diff
@@ -0,0 +1,1425 @@
...
+types may be used to define precise-width types in `<stdint.h>` or equivalent.
+
+Only these standard C++ fundamental numeric types should be used:
+
+* `bool`
+* `char` (7-bit ASCII character)
+* `int` (32 bits, signed)
+* `unsigned` (32 bits, unsigned)
+* `size_t` (pointer size, 32 bits, unsigned)
+* `ptrdiff_t` (pointer difference size, 32 bits, signed)
```
Comment (Jebbs) [inline CodingStandard.md:252]: Just a small nitpick here: the sizes of `size_t` and `ptrdiff_t` are architecture dependent. They can be 32 or 64 bits depending on the architecture you are compiling for.

File: CodingStandard.md [hunk 26]
```diff
@@ -0,0 +1,1425 @@
...
+the compiler implementer to choose or make it a configurable option.  Do not
+rely on that choice.  Instead, use `int8_t`/`uint8_t` and use `char` only if
+referring to a character in a plain C-style string.
+
+Do not use `long double`.
+
+
+**Mutable**
+
+Using the mutable keyword requires justification.
```
Comment (Jebbs) [inline CodingStandard.md]: I think using `mutable` should be prohibited. It breaks the const contract and is worse than using a `const_cast` in my opinion.

I would actually prefer `const_cast` because it is more explicit, but that would also require justification.

File: CodingStandard.md [hunk 27]
```diff
@@ -0,0 +1,1425 @@
...
+Except for dynamic_cast, use of Run-Time Type Identification (RTTI) such as
+the typeid operator requires justification.
+
+**Friend**
+
+Using the friend keyword requires justification.
+
+**Protected**
+
+Using the protected keyword requires justification.
```
Comment (Jebbs) [inline CodingStandard.md]: Why does this require special justification?

File: CodingStandard.md [hunk 28]
```diff
@@ -0,0 +1,1425 @@
...
+
+Referencing objects within a constructor that are not member variables
+or arguments to that constructor requires justification showing that these
+objects have been constructed and initialized prior to the execution of the
+constructor.
+
+**Macros**
+
+Use of C preprocessor function-macros requires justification. Use of
+inline functions are preferred.
```
Comment (Jebbs) [inline CodingStandard.md:308]: Is there anything we would want to do with macros that we can't do with inline functions? I think macros should be avoided.

File: CodingStandard.md [hunk 29]
```diff
@@ -0,0 +1,1425 @@
...
+
+Software Design Standards
+================================================================================
+
+**Software Design Constraints**
+
+- Global data must be documented in the design.
+- Global data must be minimized.
+- Bi-directional dependencies between software elements must be justified.
+- Exceptions and assertions used for control flow must be justified. Exception:
```
Comment (Jebbs) [inline CodingStandard.md]: I think these should be prohibited. This sounds worse than using a `goto`.

File: CodingStandard.md [hunk 30]
```diff
@@ -0,0 +1,1425 @@
...
+  device-independent interface to hardware services.
+
+**Object-Oriented Software Design Guidelines**
+
+- Where appropriate, make use of proven design patterns. If a design pattern
+  adds complexity, make sure the benefit outweighs the added complexity.
+- Use of interfaces as a means to decrease coupling is recommended; allow
+  clients to depend on interfaces to software elements, not on their
+  implementations.
+- Composition/aggregation is recommended instead of implementation inheritance.
```
Comment (Jebbs) [inline CodingStandard.md:349]: Why is it recommended? @RyanBerge I think you started to explain this to me, but I either forgot or you didn't have a chance to finish completely (or both).

File: CodingStandard.md [hunk 31]
```diff
@@ -0,0 +1,1425 @@
...
+code base should be justified or resolved as soon as possible.
+
+**File Naming**
+
+C++ code should adhere to these naming conventions:
+
+  - All source code filenames should be lower case.
+
+  - C++ files should use the .cpp and .h extensions (.hpp extension
+    should not be used).
```
Comment (Jebbs) [inline CodingStandard.md:372]: What's wrong with the .hpp extension?

File: CodingStandard.md (hunk 11 above)
Comment (Jebbs) [reply CodingStandard.md]: I don't like this idea. I would much rather use directories to organize everything with the top level directories being the package name. I would much prefer something like `#include <nav/localize.h>` over `#include <nav_localize.h>` and have our directory structure kind of like this:
```
├─ include/
│  └─ nav/
│     └─ localize.h
└─ src/
    └─ nav/
        └─ localize.cpp
```

File: CodingStandard.md (hunk 12 above)
Comment (Jebbs) [reply CodingStandard.md]: Just going to throw out my own naming convention as a different option or a source of ideas/conversation:

- Pre-Processor Macros: `UPPER_CASE`
  - Note: this is only for special defines, which are never referenced outside of macro usage (header guards or version specifiers only). In my opinion, the only macro that should ever be used in code is `assert`.
- Class, Struct, Union, and Enum Types: `CamelCase`
- Functions and Class Methods: `verbCamelCase`
- Compile-time constants (const, constexpr, enumerators, etc.): `UPPER_CASE`
  - Distinguished from macros by their usage in code.
- Variables and Function Parameters: `nounCamelCase`
- constexpr Methods and Functions: `verbCamelCase`
- public member variables: `nounCamelCase`
- private/protected member variables: `m_nounCamelCase`
- Const member variables: `UPPER_CASE`
  - Distinguished from macros and compile-time constants by their scope.
- Namespaces (one word) : `lower`

I haven't used templates enough in C++ to have an opinion on their naming scheme.

File: CodingStandard.md [hunk 32]
```diff
@@ -0,0 +1,1425 @@
...
+  - Template type parameters: `CamelCase`
+  - Template value parameters: `kCamelCase`
+  - Namespaces (4 characters max) : `lower_case`
+
+
+- Get methods should be named for the variable they return, and set methods
+  should be the variable name prefixed with `set_`.
+
+- Abstract base classes (interfaces) should be prefixed with I, as in
+  `IContainer`.
```
Comment (Jebbs) [inline CodingStandard.md:409]: I personally hate this convention, but I see the value in it.

I would rather have the interface documented as such and then name it normally, but that's just me.

File: CodingStandard.md (hunk 14 above)
Comment (Jebbs) [reply CodingStandard.md:440]: I prefer header guards, but not enough to say I wouldn't use `#pragma once`.

File: CodingStandard.md (hunk 15 above)
Comment (Jebbs) [reply CodingStandard.md:515]: I didn't know you could use `using` like that.

I'm with @collinsc on his views of `using` and namespaces. That we should prohibit.

edit: turns out it's there, so never mind.

File: CodingStandard.md (hunk 16 above)
Comment (Jebbs) [reply CodingStandard.md]: I've always used 80 myself.

The best argument I've heard for still using 80 is that humans read better vertically than they do horizontally, so keeping the reading area as narrow as possible helps with readability. It also let's you view two files side by side without having to scroll horizontally, which is nice.

Both of these would still be achievable with 90, but you just *barely* have enough space to view two files side by side in VS Code (my preferred editor), so I would still vote for 80.

File: CodingStandard.md [hunk 33]
```diff
@@ -0,0 +1,1425 @@
...
+
+**Special Member Functions**
+
+The special member functions of a class (constructors, copy/move constructors,
+destructors, and copy/move assignment operators) should be declared in the
+class definition and defined below the class definition in the header file if
+they are inline, or in the source file otherwise.
+
+Undesired special member functions should be marked `= delete;` inside the
+public section of the class definition (rather than making them `private`).
```
Comment (Jebbs) [inline CodingStandard.md:613]: That's some crazy voodoo. What kind of error message does one get if they dry to use a deleted function?

File: CodingStandard.md [hunk 34]
```diff
@@ -0,0 +1,1425 @@
...
+
+Except for simple struct types, default versions of all special member functions
+should be explicitly declared and defined.  Declare them in the class definition
+and define them using `= default;` in the source file.
```
Comment (Jebbs) [inline CodingStandard.md:617]: I'll have to look this up. I don't think I quite understand the problem it's trying to solve.

File: CodingStandard.md [hunk 35]
```diff
@@ -0,0 +1,1425 @@
...
+int x   = initialize_x();    // OKAY
+Foo foo = initialize_foo();  // OKAY
+
+int x;
+Foo foo;
+util::tie(x, foo) = get_tuple_of_x_and_foo();    // OKAY, if Foo isn't too large
+
+int x;
+Foo foo;
+initialize_x_and_foo(x, foo);    // OKAY, but prefer return values
```
Comment (Jebbs) [inline CodingStandard.md:710]: I don't think this should be OK.

File: CodingStandard.md [hunk 36]
```diff
@@ -0,0 +1,1425 @@
...
+```
+
+Single argument constructors and conversion functions should be marked
+`explicit` unless they are intended to participate in implicit conversions.  In
+that case, provide rationale in the conversion function routine header.
+
+**Variable Initialization**
+
+Prefer C++11 uniform initialization syntax (curly braces) for initializing 
+variables.
```
Comment (Jebbs) [inline CodingStandard.md:684]: Is using curly braces really the standard these days? That syntax bothers me a little. What is the issue with using constructors explicitly?

Maybe it is just me using languages other than C++ lately, but I would prefer something like:
`auto foo = Foo();` or `auto foo = Foo(1,2);`

Also, is something like `int x = 1;`fine, or does it have to be `int x = {1};` or `int x {1};`?

File: CodingStandard.md [hunk 37]
```diff
@@ -0,0 +1,1425 @@
...
+
+Run-time conditional checks (assertions) should be used liberally to help
+expose software bugs in development and production builds.
+
+The standard `assert(cond)` macro from `<cassert>` prints to standard output and
+aborts the system if `cond` evaluates false.  When running on target hardware,
+standard output will not be available, so `assert(cond)` should not be used.
+
+Instead, a utility `ASSERT` macro should be created that is compatible with the
+target device, and it should be used in place of `assert(cond)`.
```
Comment (Jebbs) [inline CodingStandard.md:853]: I don't agree with this. I would rather use `assert(cond)` everywhere, and redefine `assert` if we need to.

I would rather see one
```c++
#ifndef NDEBUG
    #define assert(expr) \
      //stuff
#endif // NDEBUG
```

Rather than many
```c++
#ifndef NDEBUG
    ASSERT(cond);
#else
    assert(cond);
#endif // NDEBUG
```

The name doesn't matter to me (`ASSERT` is fine) as long as the same name is everywhere. I would also be OK with this:
```c++
#ifndef NDEBUG
    #define ASSERT(expr) \
      //stuff
#else
    #define ASSERT(expr) assert(expr)
#endif // NDEBUG
```
And then just use `ASSERT` everywhere.

File: CodingStandard.md (hunk 21 above)
Comment (Jebbs) [reply CodingStandard.md:1019]: I'm fine with the way it is outlined here. Single class, multiple interface inheritance is built-in to basically every language that isn't C++.

File: CodingStandard.md (hunk 23 above)
Comment (Jebbs) [reply CodingStandard.md:1360]: If ROS already has a set of defined error codes, I vote for using them so we don't have to think.

**Review by Jebbs — COMMENTED** [review]
I reviewed the parts I'm familiar with/have an opinion on, but there are still parts that I need to do some more research on before I can comment further.

File: CodingStandard.md (hunk 3 above)
Comment (RyanBerge) [reply CodingStandard.md]: I like that suggestion.

File: CodingStandard.md (hunk 4 above)
Comment (RyanBerge) [reply CodingStandard.md]: Attribute tags are part of the C++17 standard, so we can use them if and only if we're compiling on C++17.

Comment (Jebbs) [reply CodingStandard.md]: Would it be possible to run them as part of a testing suite? As in, tests are compiled with C++17 features(where they make sense, like this one), and the actual binary on the robot is C++14?

File: CodingStandard.md (hunk 26 above)
Comment (RyanBerge) [reply CodingStandard.md]: I'm fine with that.

File: CodingStandard.md (hunk 27 above)
Comment (RyanBerge) [reply CodingStandard.md]: My understanding is that it has to do with not really being a fan of implementation inheritance in general.  I'm not sure why this was listed separately.

File: CodingStandard.md (hunk 30 above)
Comment (RyanBerge) [reply CodingStandard.md:349]: `Parent` is a base class and provides the public method `Foo()`.  `Child` is a class derived from `Parent` and provides additional functionality.  A new programmer comes along and needs to add something to `Child`.  He carelessly implements the method `Foo()` without closely examining the parent class.  `Foo()` now has a non-virtual separate implementation between `Parent` and `Child` of the same method and the compiler sees nothing wrong with this.

The resulting behavior creates a problematic bug that our new programmer takes two weeks to find.

Basically, with inheritance, methods and data get "hidden" away from our eyes, living in some other file.  With composition, the functionality and data is much easier to see.  Implementation inheritance should only be used when it is in some way **superior** to composition.

File: CodingStandard.md (hunk 31 above)
Comment (RyanBerge) [reply CodingStandard.md:372]: Nothing, I think, except for naming consistency.

File: CodingStandard.md (hunk 11 above)
Comment (RyanBerge) [reply CodingStandard.md]: Either way.  Our build system sort of ignored our file structure, so we would write `#include "snsr_subsystem.h"` instead of `#include "system/snsr/subsystem.h"`, even though that's where that file actually existed, so we needed a better identifier.

File: CodingStandard.md (hunk 33 above)
Comment (RyanBerge) [reply CodingStandard.md:613]: It's for implicit compiler-provided functions like default constructors and destructors:

```c++
class MyClass
{
    public MyClass() = delete;
    explicit public MyClass(int x);
    public MyClass(int x, int y);
    public ~MyClass() = default;
}
```

The compiler will give you an error if you instantiate `MyClass obj;`

File: CodingStandard.md (hunk 34 above)
Comment (RyanBerge) [reply CodingStandard.md:617]: It doesn't change anything in the binary file; it just shows that you left out the destructor on purpose because you wanted to use the default behavior.

File: CodingStandard.md (hunk 36 above)
Comment (RyanBerge) [reply CodingStandard.md:684]: Brace-initialization prohibits the compiler to perform implicit conversions to match constructor argument types.  <a href="https://stackoverflow.com/questions/18222926/why-is-list-initialization-using-curly-braces-better-than-the-alternatives">See this SO question for details.</a>

File: CodingStandard.md (hunk 37 above)
Comment (RyanBerge) [reply CodingStandard.md:853]: I just meant that we should have a special defined `ASSERT` that we use instead of the default.
